Hitting a release assertion in the WebExtension constructor if the resource URL doesn't contain a trailing slash

elijahsawyers opened the pull request:
#### e02f981365edfa06e7ad71c3316c9470e7ad7a30
<pre>
Hitting a release assertion in the WebExtension constructor if the resource URL doesn&apos;t contain a trailing slash
<a href="https://bugs.webkit.org/show_bug.cgi?id=286768">https://bugs.webkit.org/show_bug.cgi?id=286768</a>
<a href="https://rdar.apple.com/143775403">rdar://143775403</a>

Reviewed by Timothy Hatcher.

This patch fixes a bug where we&apos;d hit a release assert when trying to create
a WebExtension object with a resource URL that doesn&apos;t contain a trailing
slash.

We should relax this constraint and add a trailing slash if the URL doesn&apos;t
have one.

Addtionally, wrote a new test to cover this.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, LoadFromDirectoryWithoutTrailingSlash)):

Canonical link: <a href="https://commits.webkit.org/289614@main">https://commits.webkit.org/289614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4bd083bc8779550f66c2ef03ef14c7b2b1b2b23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87423 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92285 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38162 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15101 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67534 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25270 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90425 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5579 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79110 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47874 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5369 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33503 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37278 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34367 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94170 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14588 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10702 "Found 1 new test failure: fast/table/paint-section-borders-without-cells-vertical-lr-rtl.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76354 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14842 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75569 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19952 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18369 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7526 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13629 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14606 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19901 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14349 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17793 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16132 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

elijahsawyers commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -105,6 +105,9 @@
+
     RELEASE_ASSERT(m_resourceBaseURL.protocolIsFile());
     RELEASE_ASSERT(m_resourceBaseURL.hasPath());
     RELEASE_ASSERT(m_resourceBaseURL.path().right(1) == "/"_s);
```
Is this assert even needed now that we're explicitly adding the slash above?

xeenon replied:
Not really. I would also move the other asserts above the path fix lines.

elijahsawyers commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
```diff
@@ -2121,6 +2121,30 @@
+    auto *extensionURL = [NSURL URLWithString:extensionURLAbsoluteStringWithoutTrailingSlash];
+    EXPECT_NOT_NULL(extensionURL);
+
+    auto extension = adoptNS([[WKWebExtension alloc] _initWithResourceBaseURL:extensionURL error:nullptr]);
```
Is there a better way to assert that the program doesn't crash (i.e. doesn't hit a `RELEASE_ASSERT`)?

As-is, if this did start failing again, would the crash make the test results hard to infer?

xeenon replied:
A crash is a strong signal the test failed, so that is good.

elijahsawyers committed:
Hitting a release assertion in the WebExtension constructor if the resource URL doesn't contain a trailing slash
https://bugs.webkit.org/show_bug.cgi?id=286768
rdar://143775403

Reviewed by Timothy Hatcher.

This patch fixes a bug where we'd hit a release assert when trying to create
a WebExtension object with a resource URL that doesn't contain a trailing
slash.

We should relax this constraint and add a trailing slash if the URL doesn't
have one.

Addtionally, wrote a new test to cover this.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, LoadFromDirectoryWithoutTrailingSlash)):

Canonical link: https://commits.webkit.org/289614@main
